pyogrio/_io.pyx

```diff
@@ -1353,16 +1353,16 @@ def ogr_open_arrow(
         # make sure layer is read from beginning
         OGR_L_ResetReading(ogr_layer)
 
-        if not OGR_L_GetArrowStream(ogr_layer, &stream, options):
-            raise RuntimeError("Failed to open ArrowArrayStream from Layer")
-
-        stream_ptr = <uintptr_t> &stream
-
         if skip_features:
-            # only supported for GDAL >= 3.8.0; have to do this after getting
+            # only supported for GDAL >= 3.8.0; have to do this before getting
             # the Arrow stream
             OGR_L_SetNextByIndex(ogr_layer, skip_features)
 
+        if not OGR_L_GetArrowStream(ogr_layer, &stream, options):
```

Comment:
Based on my testing when implementing this, this line needed to come before `OGR_L_SetNextByIndex` or it didn't set the index properly.

Reply:
I apparently forgot to hit "comment" to save a comment about it.

It seems that one specific case in the test to filter using a sql statement + skip_features doesn't work for the `use_arrow=True` gdal >= 3.8 code path: the test where 1 row is returned by the sql statement and `skip_features=1`, so the result should be 0 rows but it is 1.

To understand better wat is happening in de gdal side for this case I had another quick look at the fix in gdal>=3.8 that enables using `skip_features` (https://github.com/OSGeo/gdal/pull/8306/files), and it basically only removes a `ResetReading` call in `OpenArrowStream` and the tests in gdal that were added in that PR also first set the `SetNextByIndex` and then do  `OpenArrowStream`.

So I tried if changing the order helped, but it gives the same result: the same test case still fails.

Reply:
Does this indicate a bug in the Arrow API in GDAL?

Reply:
I need to have another look at the gdal code to be sure, but in the failing test the `LIMIT` clause is used in the SQL statement, which the SQL equivalent of `max_features`, while `OFFSET` is equivalent of `skip_features`. Because it is a bit weird from a usage perspective to apply windowing in two ways simultaneously, I rewrote the SQL statement so it still returns 1 row, but without using a windowing function, and then the test passes.

So I think it is not (fully) supported to use both windowing clauses in the SQL statement and `skip_features`/`max_features`, but I need to test some more and read some more gdal to be sure.

I also saw that the gdal PR above also includes an optimization for GPKG to implement `skip_features` and `max_features` by appending LIMIT ... OFFSET ... to the SQL query as such, so for GPKG I'm quite sure this combo will give issues, but some more testing needed there as well as I rather expected an SQL error than a wrong result.

Reply:
I did some more tests... Not really practical without a local gdal 3.8 version, but I think I know more.

There seems to only be a minor difference in support of this feature in gdal 3.8 versus before: only whether you can put the setNextByID before OpenArrowStream or not.

The main problem is as I mentioned above that for OGRSQL sql statements (SQLITE dialect seems ok), that you cannot combine an OFFSET clause with skip_fearures. If you do, you get a wrong result.

Do for now I changed the code so that sql statements take the  local filtering code path.

I think its best to post an issue for gdal, but is a bit of a pain to create a reproducable code sample for this without using pyogrio...

Reply:
I reported the issue in the gdal issue tracker and has already been solved: [#8585](https://github.com/OSGeo/gdal/issues/8585)

It also got backported to 3.7... so will be fixed in >= 3.7.3 (once it is released).